src/ray/gcs/gcs_server/gcs_kv_manager.h

```diff
@@ -14,43 +14,135 @@
 
 #pragma once
 #include <memory>
+
+#include "absl/container/btree_map.h"
+#include "absl/synchronization/mutex.h"
 #include "ray/gcs/redis_client.h"
 #include "ray/gcs/store_client/redis_store_client.h"
 #include "ray/rpc/gcs_server/gcs_rpc_server.h"
 
 namespace ray {
 namespace gcs {
 
+/// \class InternalKVInterface
+/// The interface for internal kv implementation. Ideally we should merge this
+/// with store client, but due to compatibility issue, we keep them separated
+/// right now.
+class InternalKVInterface {
+ public:
+  /// Get the value associated with `key`.
+  ///
+  /// \param key The key to fetch.
+  /// \param callback Callback function.
+  virtual void Get(const std::string &key,
+                   std::function<void(std::optional<std::string>)> callback) = 0;
+
+  /// Associate a key with the specified value.
+  ///
+  /// \param key The key for the pair.
+  /// \param value The value for the pair.
+  /// \param overwrite Whether to overwrite existing values. Otherwise, the update
+  ///   will be ignored.
+  /// \param callback Callback function.
+  virtual void Put(const std::string &key, const std::string &value, bool overwrite,
+                   std::function<void(bool)> callback) = 0;
+
+  /// Delete the key from the store.
+  ///
+  /// \param key The key to be deleted.
+  /// \param callback Callback function.
+  virtual void Del(const std::string &key, std::function<void(bool)> callback) = 0;
+
+  /// Check whether the key exists in the store.
+  ///
+  /// \param key The key to be checked.
+  /// \param callback Callback function.
+  virtual void Exists(const std::string &key, std::function<void(bool)> callback) = 0;
+
+  /// Get the keys for a given prefix.
+  ///
+  /// \param prefix The prefix to be scaned.
+  /// \param callback Callback function.
+  virtual void Keys(const std::string &prefix,
+                    std::function<void(std::vector<std::string>)> callback) = 0;
+
+  virtual ~InternalKVInterface(){};
+};
+
+class RedisInternalKV : public InternalKVInterface {
+ public:
+  RedisInternalKV(RedisClient *redis_client) : redis_client_(redis_client) {}
+
+  void Get(const std::string &key,
+           std::function<void(std::optional<std::string>)> callback) override;
+
+  void Put(const std::string &key, const std::string &value, bool overwrite,
+           std::function<void(bool)> callback) override;
+
+  void Del(const std::string &key, std::function<void(bool)> callback) override;
+
+  void Exists(const std::string &key, std::function<void(bool)> callback) override;
+
+  void Keys(const std::string &prefix,
+            std::function<void(std::vector<std::string>)> callback) override;
+
+ private:
+  RedisClient *redis_client_;
+};
+
+class MemoryInternalKV : public InternalKVInterface {
+ public:
+  MemoryInternalKV(instrumented_io_context *io_context) : io_context_(io_context) {
```

Comment:
```suggestion
  MemoryInternalKV(instrumented_io_context &io_context) : io_context_(io_context) {
```

Should this be a reference?